src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,225 @@
+        /// </summary>
+        public static IntPtr CreateDirectoryHandle(string path)
+        {
+            IntPtr handle = Interop.Kernel32.CreateFile_IntPtr(
+                path,
+                Interop.Kernel32.FileOperations.FILE_LIST_DIRECTORY,
+                FileShare.ReadWrite | FileShare.Delete,
+                FileMode.Open,
+                Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle == IntPtr.Zero || handle == (IntPtr)(-1))
+            {
+                // Historically we throw directory not found rather than file not found
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            if (Interlocked.Exchange(ref _state, 1) == 0 && _threadId == Environment.CurrentManagedThreadId)
+            {
+                InitEnumeration();
+                return this;
+            }
+
+            FindEnumerable<TResult> clone = new FindEnumerable<TResult>(_originalUserPath, _originalFullPath, _transform, _predicate, _recursive);
+            clone.InitEnumeration();
+            return clone;
+        }
+
+        private void InitEnumeration()
+        {
+            // We want to delay creating any handles until the enumerator is actually created to
+            // avoid keeping handles open any longer than needed. Once we actually start enumerating
+            // it's fair game to stash handles until we've finished processing the given directory.
+
+            _directoryHandle = CreateDirectoryHandle(_originalFullPath);
+            _currentPath = _originalFullPath;
+            _buffer = ArrayPool<byte>.Shared.Rent(4096);
+            _pinnedBuffer = GCHandle.Alloc(_buffer, GCHandleType.Pinned);
+            if (_recursive)
+                _pending = new Queue<(IntPtr, string)>();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public TResult Current
+        {
+            get
+            {
+                RawFindData findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                return _transform(ref findData);
+            }
+        }
+
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
+        {
+            if (_lastEntryFound)
+                return false;
+
+            RawFindData findData = default;
+            do
+            {
+                FindNextFile();
+                if (!_lastEntryFound && _info != null)
+                {
+                    // If needed, stash any subdirectories to process later
+                    if (_pending != null && (_info->FileAttributes & FileAttributes.Directory) != 0
```

Comment:
_pending would only be null if this object has been Disposed. Are you attempting to preotect against that everywhere?

Reply:
_pending is also used to track recursive- now that I've refactored I'll just use _recursive